### ISSUE TITLE
Mismatch num_samples for Torch and Tensorflow implementation IG

### ISSUE THREAD

**Issue by jumelet** [suinleelab/path_explain#8]
I'm currently testing the Integrated Gradients implementation on a torch and tensorflow version of a Huggingface model, and noticed the attributions I got with the same configuration are slightly different.

I tested this with `use_expectations=False`, so for the Integrated Gradients and not the Expected Gradients.

It turns out that the `num_samples` argument behaves slightly different: for the torch model `_get_samples_input` returns an interpolated baseline that is of size `num_samples + 1`, which is happening at this point due to the +1:

https://github.com/suinleelab/path_explain/blob/b567945fe02ab8a6e6de675c8367edd45287b243/path_explain/explainers/path_explainer_torch.py#L102-L103

The `_sample_alphas` method of `PathExplainerTF`, on the other hand, returns an interpolation tensor of length `num_samples`, which causes the mismatch. 

When passing a `num_samples` to the torch implementation that is 1 lower instead the attributions of both methods are exactly the same.

I'm not sure which of these two is more "correct", but I think it would be good if `num_samples` acts the same for both implementations. One could even argue that in both cases we should return `num_samples+2` points, given that the interpolation also includes the input + baseline itself; in that case the number of intermediate points based on which we'd be computing the attributions would actually be equal to `num_samples`.

**Comment by jjanizek** [comment]
> I think it would be good if num_samples acts the same for both implementations.

Yes, i think that's a great point and the two implementations should have the same behavior. I think I'd tend towards having `num_samples` represent the total number of gradient calls the explainer is doing, so I'll probably change the `PathExplainerTorch` implementation to more closely match the `_sample_alphas` method of `PathExplainerTF`

**Comment by jjanizek** [comment]
Would you be willing to send me a snippet of the code for the HuggingFace model you've been looking at in TF and PyTorch? I was planning to do some maintenance on the explainers this week, and that seems like a very good test for ensuring consistency of behavior between the two explainers

**Comment by jumelet** [comment]
Yes, no problem at all, I'll quickly set up a code snippet

**Comment by jumelet** [comment]
I've been testing your code on a causal language model, DistilGPT-2, on a simple input sentence ("It's raining cats and dogs"). The attributions that are computed here are for "dogs" based on the prefix "It's raining cats and".

## Torch
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from path_explain import PathExplainerTorch, text_plot

model = AutoModelForCausalLM.from_pretrained("distilgpt2")
tokenizer = AutoTokenizer.from_pretrained("distilgpt2")

def embedding_model(batch_ids):
    batch_embedding = model.transformer.wte.weight[[batch_ids]]
    return batch_embedding

def prediction_model(batch_embedding):
    logits = model(inputs_embeds=batch_embedding).logits[:, -1]
    probs = torch.log_softmax(logits, dim=-1)
    
    return probs

sen = "It is raining cats and"

batch_ids = tokenizer([sen])['input_ids']
batch_embedding = embedding_model(batch_ids)
baseline_embedding = torch.zeros_like(batch_embedding)
output_idx = torch.tensor([tokenizer.convert_tokens_to_ids(f"Ġdogs")])

explainer = PathExplainerTorch(prediction_model)

attributions = explainer.attributions(
    batch_embedding,
    baseline_embedding,
    output_indices=output_idx,
    num_samples=9,
    use_expectation=False,
)

sum_attr = attributions.squeeze().detach().numpy().sum(-1)  # Sum over embedding dim

text_plot(
    sen.split(),
    sum_attr,
)
```

## Tensorflow
```python
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

from scipy.special import log_softmax
from transformers import TFAutoModelForCausalLM, AutoTokenizer

from path_explain import EmbeddingExplainerTF, text_plot

tf_model = TFAutoModelForCausalLM.from_pretrained("distilgpt2")
tf_tokenizer = AutoTokenizer.from_pretrained("distilgpt2")

def embedding_model(batch_ids):
    batch_embedding = tf_model.transformer.wte(batch_ids)
    return batch_embedding

def prediction_model(batch_embedding):
    if isinstance(batch_embedding, np.ndarray):
        batch_embedding = tf.convert_to_tensor(batch_embedding)
    logits = tf_model(None, inputs_embeds=batch_embedding).logits[:, -1]
    probs = tf.nn.log_softmax(logits, axis=-1)
    
    return probs

sen = "It is raining cats and"

batch_ids = tf_tokenizer([sen])['input_ids']
batch_embedding = embedding_model(batch_ids)

baseline_embedding = np.zeros_like(batch_embedding)

output_idx = tf_tokenizer.convert_tokens_to_ids("Ġdogs")

explainer = EmbeddingExplainerTF(prediction_model)
attributions = explainer.attributions(
    batch_embedding,
    baseline_embedding,
    output_indices=output_idx,
    num_samples=10,
    batch_size=1,
    use_expectation=False,
)

text_plot(
    sen.split(),
    attributions[0],
)
```

**Comment by jjanizek** [comment]
What version of transformers and pytorch are you using? I'm actually getting a weird issue where the gradients with respect to the embedding are really unreasonable in PyTorch while looking totally correct in TF

**Comment by jumelet** [comment]
Ah that's odd, I'm currently on torch v.1.8.0, tensorflow v2.5.0, and transformers v4.8.2.

**Comment by jjanizek** [comment]
Great -- apparently there's some issue with torch v.1.6 and transformers v4.8, now that I've upgraded to torch v.1.9 the gradients appear to be of the correct magnitude and consistent with what I've seen in tensorflow (all our NLP experiments were initially done in TF, which is why we never caught this).

I think you're correct that for the sake of consistency, the "+1" should be removed from the PyTorch explainer so that it instead looks like this:
`scaled_inputs = [reference_tensor + (float(i)/num_samples)*(input_expand - reference_tensor) \ 
                  for i in range(0,num_samples)] `

Again, feel free to make a pull request if you'd like, and nice catch!
